pkg/maglev/maglev.go

```diff
@@ -64,16 +84,20 @@ func getOffsetAndSkip(backend string, m uint64) (uint64, uint64) {
 
 func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
 	var wg sync.WaitGroup
-	bCount := len(backends)
-	perm := make([]uint64, bCount*int(m))
 
 	// The idea is to split the calculation into batches so that they can be
-	// concurrently executed.  We limit the number of concurrent goroutines to
+	// concurrently executed. We limit the number of concurrent goroutines to
 	// the number of available CPU cores. This is because the calculation does
 	// not block and is completely CPU-bound. Therefore, adding more goroutines
 	// would result into an overhead (allocation of stackframes, stress on
 	// scheduling, etc) instead of a performance gain.
 
+	bCount := len(backends)
+	if b := uint64(bCount); b > maxBackends {
+		maxBackends = b // Update the max so we don't have to allocate again.
+		permutation = make([]uint64, maxBackends*m)
```

Comment:
I'd not update the pre-allocated permutation slice in this case. Otherwise, there is no way for users to opt-out from the pre-allocation which might be the case for memory-tight systems.

Reply:
Is that really the case? For memory-tight systems, we can recommend to users to set the max backends to 0, if needed. Then this code will only allocate memory for how many backends the user has and only when they use Maglev. 

For example, the user adds 5 backends. The code will reallocate the slice to accommodate 5 backends. If the user delete 4 backends, the slice is still allocated for 5 backends.

Unless memory-tight users add many backends and then quickly remove them, I don't really see the need to account for such a use-case.

Is it worth accounting for memory-tight users when using Maglev? I presume that users who are going to use Maglev won't be memory-tight. What do you think?

Reply:
> Is that really the case? For memory-tight systems, we can recommend to users to set the max backends to 0, if needed. Then this code will only allocate memory for how many backends the user has and only when they use Maglev.

My point is that even users set it to `0`, the GC won't collect that slice after we have finished with Maglev lookup calculation. If the users want this slice to be kept, then they can set the param large enough to accommodate all backends.

> Is it worth accounting for memory-tight users when using Maglev? I presume that users who are going to use Maglev won't be memory-tight and won't have the use-case outlined above. What do you think?

This is a good point. WDYT @borkmann ?

Reply:
Do you have some ballpark numbers wrt how much memory we're talking about for the heuristic?

Reply:
It depends on backends count. The mem which won't be returned to OS is `|backends| * M * sizeof(uint64)`. If there is an outlier service with many backends and rarely updated, then this could lead to wasted memory, and there is no way to opt-out from this. E.g. with M=16381 and |backends| = 100, the perm slice is~12.5MB.

Reply:
Couldn't we derive a heuristic based on system mem? e.g. if the node has >=8GB I don't think this matters overly much in the big picture.

Reply:
Probably we could based on sys mem and M.

Reply:
Can we simply say that any node running with less than 8GB will not allocate the perm slice ahead of time and nodes with more than 8GB will use 5000 backends as the starting point? If they exceed 5000 backends, then the already existing logic will account for that and increase the ceiling?

Reply:
I'd suggest if sys mem > 8GB, then do the initial pre-allocation of perm slice of the `(M / 100) * M` size (`M / 100` suggest max backends count because of the property `|backends| * 100 < M`). Once we detect that the perm slice is too small, we pre-allocate a bigger one as it is implemented now. This gives us the following mem pressure with different M:
- 251 : 0.004806594848632812 MB
- 509 : 0.019766311645507812 MB
- 1021 : 0.07953193664550783 MB
- 2039 : 0.3171936798095703 MB
- 4093 : 1.2781256866455077 MB
- 8191 : 5.118750076293945 MB
- 16381 : 20.472500686645507 MB
- 32749 : 81.82502754211426 MB
- 65521 : 327.5300171661377 MB
- 131071 : 1310.700000076294 MB

Reply:
Implemented the above. Thanks @brb